#!/bin/bash

#--------------- EXECUTABLE ---------------

#STEP 8

#this script renames individual nextgen sequence files from NCBI given name when downloaded to sequential sample names (sample0.fq, sample1.fq, sample2.fq, etc.)
#using the sample naming key so Stacks assembly pipeline and other scripts will play nice with samples
#
# input: readyforrenaming_XXX.XXX.gz file of nextgen genetic sequence data - low quality reads and common adapter sequences removed
#			input files are stored in directory at <storagenode>/<user>/bioprj_<NCBIbioprj#>_<NCBIspeciesname>/3_adapterrmvd_reads
#			example: /mnt/scratch/rhtoczyd/bioprj_PRJNA524160_Impatiens-capensis/3_adapterrmvd_reads/readyforrenaming_SRR8625100.fastq.gz
#		 master_keys/samplenamingkey-XXXX.txt file that contains original NCBI sample name and new bioinformatics/Stacks sample name
#
# output: sample0.XX.gz, sample1.XX.gz files of nextgen genetic sequence data - ready for Stacks pipeline (starting with ustacks)
#			output files are stored in directory at <storagenode>/<user>/bioprj_<NCBIbioprj#>_<NCBIspeciesname>/samples
#			example: /mnt/scratch/rhtoczyd/bioprj_PRJNA524160_Impatiens-capensis/samples/sample0.fq.gz

#This script uses files generated by R script: prepare_to_download_sequence_data.R to rename files from NCBI SRA number to sample0, sample1, etc. in prep for Stacks pipeline
#It also puts a popmap and sample naming key in each /samples directory on storage node for each dataset (bioprj/species combo)


#define variables
storagenode=/mnt/scratch/$USER #path to main node to where fastq files live

list_of_datasets=list-coral.txt #name of file that contains the name of all of the directories that we want to process

samplenamekeydir=/mnt/home/rhtoczyd/divdiv/data/all_samplenamekeys #file path where sample name keys live, that specify how to rename sequence files from SRRXXXXX to sample0, sample1, etc. for Stacks pipeline

popmapdir=/mnt/home/rhtoczyd/divdiv/data/bioinformatics/popmaps #file path to where popmaps live

#for each sample directory of a bioprj/species combo dataset in the list provided, rename fastq files

while read dataset
do 	

	#define dir name that seq. files live inside
	run_name=$(echo $dataset | cut -d " " -f1 | cut -d "." -f1)
	echo dataset is $dataset
	echo run_name is $run_name
	
	#define path to where sequence data with low quality reads and common adapters removed live for this dataset
	#and where renamed reads should be written to, ready for Stacks
	finalreadsdir=$storagenode/$run_name/4_cleanandtrimmedtoconsistentlength_reads
	samp_storage_dir=$storagenode/$run_name/samples
	if [ ! -d $samp_storage_dir ]; then mkdir $samp_storage_dir; fi
	
	#for each sequence file in the cleaned reads dir,
	for seqfile in $finalreadsdir/*fq.gz
	do
		#get NCBI sequence ID (alone as a variable)
		echo seqfile is $seqfile
		RUN_ACC_ID=${seqfile//$finalreadsdir/}
		RUN_ACC_ID=$(echo $RUN_ACC_ID | cut -d/ -f2)
		RUN_ACC_ID=$(echo $RUN_ACC_ID | cut -d. -f1)
		RUN_ACC_ID=$(echo $RUN_ACC_ID | cut -d "_" -f3)
		echo RUN_ACC_ID is $RUN_ACC_ID
		
		#find NCBI sequence ID in sample naming key and get new name for sample
		new_sample_name=$(grep $RUN_ACC_ID ${samplenamekeydir}/samplenamekey-${run_name}.txt | cut -d$'\t' -f3)
		echo new_sample_name is $new_sample_name
		
		#copy sample from cleaned_reads/ to samples/ and rename it in new location (at which point samples are ready for Stacks)
		cp $seqfile $samp_storage_dir/$new_sample_name.fq.gz
		echo seqfile is $seqfile
		
		echo I renamed sample $RUN_ACC_ID to $new_sample_name
	done
	
	
	#define current name of popmap for dataset we're processing
	current_popmap_file_name=popmap_${run_name}.txt
	echo current_popmap_file_name is $current_popmap_file_name
	
	#copy popmap from popmap folder (where R code generated all popmaps) to dir for this dataset ($storagenode/$run_name)
	#and rename from long name with bioprj/species info to just popmap (so Stacks code will run correctly) 
	cp $popmapdir/$current_popmap_file_name $storagenode/$run_name/popmap
	
	#and store a copy of sample naming key just for this dataset in dir too, insurance policy/good bookkeeping
	cp ${samplenamekeydir}/samplenamekey-${run_name}.txt $storagenode/$run_name/samplenamekey.txt

done < ../../master_keys/$list_of_datasets

echo DONE